File: iosApp/iosApp/Views/RefreshableViewModel.swift
import Shared
import SwiftUI

@Observable
open class RefreshableViewModel {
    var isRefreshing = false
    private let alertState = AlertStateInjector()
    
    func task<T>(_ closure: () async throws -> T) async -> T? {
        do {
            self.isRefreshing = true
            let result = try await closure()
            self.isRefreshing = false
            return result
        } catch {
            self.alertState.alertState.show(message: error.localizedDescription)
            return nil
        }
    }
    
    func spawnTask(_ closure: @escaping () async throws -> Void) {
        Task {
            await self.task(closure)
        }
    }
}

File: composeApp/src/androidMain/kotlin/com/github/aeoliux/violet/app/RefreshableViewModel.kt
package com.github.aeoliux.violet.app

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.github.aeoliux.violet.repositories.AlertState
import com.github.aeoliux.violet.repositories.AlertStateInjector
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

open class RefreshableViewModel(): ViewModel() {
    private var _isRefreshing = MutableStateFlow(false)
    val isRefreshing get() = _isRefreshing.asStateFlow()

    fun task(closure: suspend () -> Unit) {
        viewModelScope.launch {
            _isRefreshing.update { true }
            closure()
            _isRefreshing.update { false }
        }
    }
}